docs/devGuide/migratingToTypeScript.md

```diff
@@ -0,0 +1,135 @@
+
+   If the dependency has its typings:
+
+   - Bundled alongside their implementation, you don't need to do anything else.
+
+   - In a `@types/<lib-name>` module, install it as a development dependency for the `core` package by navigating to `packages/core` and running `npm i -D @types/<lib-name>`.
+
+     After installing all necessary typings, delete the generated `packages/core/package-lock.json`, navigate back to the root directory, and run `npm run setup`.
+
+     <box type="info">
+
+     It's recommended to try to match the `@types/<lib-name>` version with the associated module. In general, try to match the major and minor version as closely as possible.
+     </box>
+
+3. **Stash the changes** (`packages/core/package.json` and root `package-lock.json`).
+   <br>
+   This will be added back during the "Adapt" commit later.
+
+### Step 2: "Rename" Commit
+
+1. **Add the paths to the `.js` files to `.gitignore` and `.eslintignore`**.
+   <br>
+   This instructs Git and ESLint to ignore the compiled JavaScript from the files that are going to be renamed.
+
+2. **Rename the files** from having `.js` extensions to `.ts` extensions.
+
+3. **Stage the changes** (the renamed files, `.gitignore`, and `.eslintignore`) for commit.
+   <br>
+   Ensure the renamed files are regarded by Git as `renamed` and not `added`.
+
+4. **Commit the staged changes** with the message `Rename <file-or-folder> to TypeScript`.
+   <br>
+   For example: `Rename core/src/html to TypeScript`.
+
+### Step 3: "Adapt" Commit
+
+1. **Start automatic compilation on change/save**. More on this at [Editing backend features](workflow.md#editing-backend-features) section in the Workflow page.
+
+2. **Change the CommonJS module import/export syntax** to TypeScript's equivalent syntax (use ES6 syntax only if possible and/or necessary).
+
+   A common error in newly renamed files comes from the way modules are imported/exported. The syntax for those can be converted according to the table below:
+
+   Case           | CommonJS                   | TypeScript equivalent                                            | ES6
+   ---------------|----------------------------|------------------------------------------------------------------|-----------------------
+   Default import | `const x = require(y)`     | `import x = require(y)`                                          | `import x from y`
+   Named import   | `const { x } = require(y)` | N/A, workaround:<br>`import z = require(y)`<br>`const { x } = z` | `import { x } from y`
+   Named export   | `module.exports = { x }`   | `export = { x }`                                                 | `export { x }`
+   Default export | `module.exports = x`       | `export = x`                                                     | `export default x`
```

Comment:
referring https://github.com/MarkBind/markbind/pull/1877#discussion_r851785006

Just one nit as such while you're still touching things up -- let's make it clear the regular `import x = require(y)` / `export = x` isn't default import export. (Maybe just give a warning, and link the typescript docs / playground)

E.g. `import x from y` actually compiles to this:

![ts](https://user-images.githubusercontent.com/3306138/163726482-119d6bfe-c446-4d2a-8dd7-0754e6fcb324.png)

---

Remember to push fixups as well! https://github.com/MarkBind/markbind/pull/1877#pullrequestreview-943841714

Reply:
Apologies if my definitions might lead to some confusion, but to clarify, the "default" import/export I'm mentioning here is rather more to the concept of importing/exporting a single (non-object) thing, such a constant, a function, a class, etc. It is why I associate `const x = require(y)`, `import x = require(y)`, and `import x from y` together.

I have also noted in the docs that, as exampled in the image above, the ES6 syntax will compile into something more than just CommonJS syntax, and thus proceed with ES6 only if possible. I guess what I'm trying to say is that while they are doing the same job at high-level (import/export a single thing), they are not identical at low-level.

I will update the docs to clarify this more.

Reply:
maybe:
- "module.exports object"
- "Entire export"? (still seems ambiguous)
- "`export` object"

Reply:
I think there is no agreeable term for the concept. Most just simply say one is to "export one thing" and the other to "export many things". There are developer-written articles (like [this](https://www.freecodecamp.org/news/node-module-exports-explained-with-javascript-export-function-examples/) and [this](https://www.sitepoint.com/understanding-module-exports-exports-node-js/)) kind of use the same definition as I wrote earlier, "export one thing" = "default export" in CommonJS context.

Reply:
I just updated the docs to clarify this, and since it is becoming too long I put the syntax reference on a separate section to be referred to from the migration steps. I opt for the simpler "export one thing" vs "export many things" to describe the different cases.

Reply:
> I think there is no agreeable term for the concept. Most just simply say one is to "export one thing" and the other to "export many things". There are developer-written articles (like [this](https://www.freecodecamp.org/news/node-module-exports-explained-with-javascript-export-function-examples/) and [this](https://www.sitepoint.com/understanding-module-exports-exports-node-js/)) kind of use the same definition as I wrote earlier, "export one thing" = "default export" in CommonJS context.

No, do refer to a canonical, or at least peer reviewed source. https://nodejs.org/api/modules.html, https://stackoverflow.com/questions/21117160/what-is-export-default-in-javascript, https://stackoverflow.com/questions/40294870/module-exports-vs-export-default-in-node-js-and-es6

There is no concept of "default import/export" in commonJs, this is a very common misconception that the first article you linked has clearly made as well. The second article refers to the vague concept of a "default value", not "default import/export" (which is a technical term formalised in es6 proposals).

This is important as its a very common misconception, one that I made when I started learning about this as well, and as you mentioned has very real implementation implications. We have contributors who aren't as familiar with module systems coming in; Developer documentation above all else should aim to be technically accurate and unambiguous.

The canonical term here would simply be `module.exports` or `exports` object (https://nodejs.org/api/modules.html).

Default exports, are in fact, closer to the concept of named exports in the context of commonJS. https://stackoverflow.com/questions/40294870/module-exports-vs-export-default-in-node-js-and-es6

> I opt for the simpler "export one thing" vs "export many things" to describe the different cases.

Sounds better, and good work here 👍 

Some minor suggestions dropped below